### DISCUSSION TITLE
How to use `tracing-flame`? Getting empty file.

### DISCUSSION THREAD

**Discussion by rjzak** [tokio-rs/tracing#2995] [Q&A]
I have a server app that I'm working on and wish to use `tracing-flame` to generate a flame chart. It just creates an empty file though.

I create the flame layer and have `let guard = flame_layer.flush_on_drop();`. But the server never exits on it's own, just at Control-C to stop it, or if I stop the service with the `service` command. Is it because the program doesn't exit on it's own that maybe the guard isn't properly dropped? The Drop should still be handled accordingly, correct? If not, how can I trigger the flame graph data to be saved to disk right before Control-C?

**Comment by mladedav** [comment]
When default handlers run, they just exit without unwinding and running drops like an unwinding panic does.

You can try using the `ctrlc` crate or another one that will allow you to handle other signals like sigterm too.

**Reply by rjzak** [reply]
Makes sense. How would I have the `ctrlc` hook, which handles forced exit, to have the flamegraph saved? An explicit `drop` of the guard?

**Reply by mladedav** [reply]
I suppose calling `flush` should be enough.
